src/libraries/System.Text.RegularExpressions/gen/UpgradeToRegexGeneratorCodeFixer.cs

```diff
@@ -103,16 +103,21 @@ private static async Task<Document> ConvertToSourceGenerator(Document document,
             }
 
             // Get the parent type declaration so that we can inspect its methods as well as check if we need to add the partial keyword.
-            TypeDeclarationSyntax? typeDeclaration = nodeToFix.Ancestors().OfType<TypeDeclarationSyntax>().FirstOrDefault();
+            SyntaxNode? typeDeclarationOrCompilationUnit = nodeToFix.Ancestors().OfType<TypeDeclarationSyntax>().FirstOrDefault();
 
-            if (typeDeclaration is null)
+            if (typeDeclarationOrCompilationUnit is null)
             {
-                return document;
+                typeDeclarationOrCompilationUnit = await nodeToFix.SyntaxTree.GetRootAsync(cancellationToken).ConfigureAwait(false);
             }
 
             // Calculate what name should be used for the generated static partial method
             string methodName = DefaultRegexMethodName;
-            ITypeSymbol? typeSymbol = semanticModel.GetDeclaredSymbol(typeDeclaration, cancellationToken) as ITypeSymbol;
+            INamedTypeSymbol? typeSymbol = typeDeclarationOrCompilationUnit switch
+            {
+                TypeDeclarationSyntax typeDeclaration => semanticModel.GetDeclaredSymbol(typeDeclaration, cancellationToken),
+                CompilationUnitSyntax compilationUnit => semanticModel.GetDeclaredSymbol(compilationUnit, cancellationToken)?.ContainingType,
+                _ => throw new InvalidOperationException("This is unreachable."),
```

Comment:
While I understand this should be unreachable, previously if we had any case where typedeclaration was null we were just returning the document unchanged, should we just continue to do the same?

Reply:
If the case where the document is returned unchanged is reachable, then I think it's fair to consider it as a bug. This basically mean we'll offer the user a code action, but the code action doesn't do anything. These cases are to be filtered before we register the code action.

Reply:
fair enough

Reply:
If it's actually unreachable, then why use a switch like this?  It could just be a ternary:
```C#
INamedTypeSymbol? typeSymbol = typeDeclarationOrCompilationUnit is TypeDeclarationSyntax typeDeclaration ?
    semanticModel.GetDeclaredSymbol(typeDeclaration, cancellationToken) :
    semanticModel.GetDeclaredSymbol((CompilationUnitSyntax)typeDeclarationOrCompilationUnit, cancellationToken)?.ContainingType;
```

Reply:
@stephentoub Thanks. I applied your suggestion in last commit.

I just used to write it that way since it's a common pattern in Roslyn. They even have a dedicated static property for it https://github.com/dotnet/roslyn/blob/7e2478f3947c1721619341d702152365ba5581ae/src/Compilers/Core/Portable/InternalUtilities/ExceptionUtilities.cs#L27-L30 :)